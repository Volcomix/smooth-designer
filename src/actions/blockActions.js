//@flow
import {
  ADD_BLOCK,
  UPDATE_BLOCK_NAME,
  UPDATE_BLOCK_SIZE,
} from '../constants/actionTypes'

type AddBlockAction = { type: typeof ADD_BLOCK }

type UpdateBlockNameAction = {
  type: typeof UPDATE_BLOCK_NAME,
  id: string,
  name: string,
}

type UpdateBlockSizeAction = {
  type: typeof UPDATE_BLOCK_SIZE,
  id: string,
  width: number,
  height: number,
}

export type BlockAction =
  | AddBlockAction
  | UpdateBlockNameAction
  | UpdateBlockSizeAction

export const addBlock = (): AddBlockAction => ({ type: ADD_BLOCK })

export const updateBlockName = (
  id: string,
  name: string,
): UpdateBlockNameAction => ({ type: UPDATE_BLOCK_NAME, id, name })

export const updateBlockSize = (
  id: string,
  width: number,
  height: number,
): UpdateBlockSizeAction => ({ type: UPDATE_BLOCK_SIZE, id, width, height })
