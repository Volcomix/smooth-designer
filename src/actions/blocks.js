//@flow
import {
  ADD_BLOCK,
  DELETE_BLOCK,
  UPDATE_BLOCK_NAME,
  UPDATE_BLOCK_SIZE,
} from '../constants/actionTypes'

type AddBlockAction = { type: typeof ADD_BLOCK }

type DeleteBlockAction = { type: typeof DELETE_BLOCK, id: string }

type UpdateBlockNameAction = {
  type: typeof UPDATE_BLOCK_NAME,
  id: string,
  name: string,
}

type UpdateBlockSizeAction = {
  type: typeof UPDATE_BLOCK_SIZE,
  id: string,
  width: number,
  height: number,
}

export type BlocksAction =
  | AddBlockAction
  | DeleteBlockAction
  | UpdateBlockNameAction
  | UpdateBlockSizeAction

export const addBlock = (): AddBlockAction => ({ type: ADD_BLOCK })

export const deleteBlock = (id: string): DeleteBlockAction => ({
  type: DELETE_BLOCK,
  id,
})

export const updateBlockName = (
  id: string,
  name: string,
): UpdateBlockNameAction => ({ type: UPDATE_BLOCK_NAME, id, name })

export const updateBlockSize = (
  id: string,
  width: number,
  height: number,
): UpdateBlockSizeAction => ({ type: UPDATE_BLOCK_SIZE, id, width, height })
