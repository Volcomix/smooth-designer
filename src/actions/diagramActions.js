//@flow
import { START_LINKING, END_LINKING } from '../constants/actionTypes'

type StartLinkingAction = {
  type: typeof START_LINKING,
  fromBlock: string,
}

type EndLinkingAction = {
  type: typeof END_LINKING,
  toBlock: string,
}

export type DiagramAction = StartLinkingAction | EndLinkingAction

export const startLinking = (fromBlock: string): StartLinkingAction => ({
  type: START_LINKING,
  fromBlock,
})

export const endLinking = (toBlock: string): EndLinkingAction => ({
  type: END_LINKING,
  toBlock,
})
