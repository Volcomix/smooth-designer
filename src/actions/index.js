//@flow
import type { BlockAction } from './blockActions'
import type { ForceAction } from './forceActions'
import type { DiagramAction } from './diagramActions'

export type Action = BlockAction | ForceAction | DiagramAction
