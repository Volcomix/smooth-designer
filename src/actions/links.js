//@flow
import {
  START_LINKING,
  UPDATE_LINKING,
  END_LINKING,
  CANCEL_LINKING,
} from '../constants/actionTypes'

type StartLinkingAction = {
  type: typeof START_LINKING,
  fromId: string,
  toMouse: { x: number, y: number },
}

type UpdateLinkingAction = {
  type: typeof UPDATE_LINKING,
  toMouse: { x: number, y: number },
}

type EndLinkingAction = {
  type: typeof END_LINKING,
  toId: string,
}

type CancelLinkingAction = {
  type: typeof CANCEL_LINKING,
}

export type LinksAction =
  | StartLinkingAction
  | UpdateLinkingAction
  | EndLinkingAction
  | CancelLinkingAction

export const startLinking = (
  fromId: string,
  toMouseX: number,
  toMouseY: number,
): StartLinkingAction => ({
  type: START_LINKING,
  fromId,
  toMouse: { x: toMouseX, y: toMouseY },
})

export const updateLinking = (
  toMouseX: number,
  toMouseY: number,
): UpdateLinkingAction => ({
  type: UPDATE_LINKING,
  toMouse: { x: toMouseX, y: toMouseY },
})

export const endLinking = (toId: string): EndLinkingAction => ({
  type: END_LINKING,
  toId,
})

export const cancelLinking = (): CancelLinkingAction => ({
  type: CANCEL_LINKING,
})
