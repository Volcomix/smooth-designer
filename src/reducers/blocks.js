//@flow
import type { Block } from '../types'
import type { Action } from '../actions'
import {
  ADD_BLOCK,
  UPDATE_BLOCK_NAME,
  UPDATE_BLOCK_SIZE,
  UPDATE_FORCE,
} from '../constants/actionTypes'

export type BlocksState = { [id: string]: Block }

const initialState: BlocksState = {}

const addBlock = state => {
  const ids: number[] = Object.keys(state).map(id => parseInt(id, 10))
  let id = 0
  if (ids.length > 0) {
    id = Math.max(...ids) + 1
  }
  return {
    ...state,
    [id]: { id: `${id}`, name: '', x: 0, y: 0, width: 0, height: 0 },
  }
}

const updateBlockName = (state, { id, name }) => ({
  ...state,
  [id]: { ...state[id], name },
})

const updateBlockSize = (state, { id, width, height }) => ({
  ...state,
  [id]: { ...state[id], width, height },
})

const updateBlocksPositions = (state, { blocks }) =>
  blocks.reduce((nextState, { id, x, y }) => {
    nextState[id] = { ...state[id], x, y }
    return nextState
  }, {})

const blocks = (
  state: BlocksState = initialState,
  action: Action,
): BlocksState => {
  switch (action.type) {
    case ADD_BLOCK:
      return addBlock(state)
    case UPDATE_BLOCK_NAME:
      return updateBlockName(state, action)
    case UPDATE_BLOCK_SIZE:
      return updateBlockSize(state, action)
    case UPDATE_FORCE:
      return updateBlocksPositions(state, action)
    default:
      return state
  }
}

export const getBlocks = (state: BlocksState): Block[] =>
  (Object.values(state): any)

export const cloneBlocks = (state: BlocksState): Block[] =>
  getBlocks(state).map(block => ({ ...block }))

export default blocks
