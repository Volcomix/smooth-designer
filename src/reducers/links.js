//@flow
import type { Link } from '../types'
import type { Action } from '../actions'
import {
  START_LINKING,
  UPDATE_LINKING,
  END_LINKING,
  CANCEL_LINKING,
} from '../constants/actionTypes'

export type LinksState = {
  +links: { [id: string]: Link },
  +linking?: {
    +fromId: string,
    +toMouse: {
      +x: number,
      +y: number,
    },
  },
}

const initialState: LinksState = { links: {} }

const startLinking = ({ links }, { fromId, toMouse }) => ({
  links,
  linking: { fromId, toMouse },
})

const updateLinking = ({ links, linking }, { toMouse }) => ({
  links,
  linking: { ...linking, toMouse },
})

const endLinking = ({ links, linking }, { toId }) => {
  if (!linking) {
    throw new Error('No linking defined.')
  }
  const ids: number[] = Object.keys(links).map(id => parseInt(id, 10))
  let id = 0
  if (ids.length > 0) {
    id = Math.max(...ids) + 1
  }
  return {
    links: {
      ...links,
      [id]: {
        id: `${id}`,
        fromId: linking.fromId,
        toId,
      },
    },
    linking,
  }
}

const cancelLinking = ({ links }) => ({
  links,
})

const links = (
  state: LinksState = initialState,
  action: Action,
): LinksState => {
  switch (action.type) {
    case START_LINKING:
      return startLinking(state, action)
    case UPDATE_LINKING:
      return updateLinking(state, action)
    case END_LINKING:
      return endLinking(state, action)
    case CANCEL_LINKING:
      return cancelLinking(state)
    default:
      return state
  }
}

export const getLinks = ({ links }: LinksState): Link[] =>
  (Object.values(links): any)

export default links
