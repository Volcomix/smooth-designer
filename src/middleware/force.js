//@flow
import {
  forceSimulation,
  forceX,
  forceY,
  forceCollide,
  forceCenter,
  forceLink,
} from 'd3-force'
import type { Middleware } from 'redux'
import type { Block } from '../types'
import type { State } from '../reducers'
import type { Action } from '../actions'
import { UPDATE_BLOCK_SIZE, END_LINKING } from '../constants/actionTypes'
import { cloneBlocks } from '../reducers/blocks'
import { getLinks } from '../reducers/links'
import { updateForce } from '../actions/force'

export const radius = ({ width, height }: Block) =>
  Math.hypot(width, height) / 2

const simulation = forceSimulation()
  .force('x', forceX().strength(0.09))
  .force('y', forceY().strength(0.16))
  .force('collision', forceCollide(radius))
  .force('center', forceCenter())
  .stop()

const computeStaticLayout = () => {
  for (
    let i = 0,
      n = Math.ceil(
        Math.log(simulation.alphaMin()) / Math.log(1 - simulation.alphaDecay()),
      );
    i < n;
    ++i
  ) {
    simulation.tick()
  }
}

const force: Middleware<State, Action> = ({
  dispatch,
  getState,
}) => next => action => {
  const result = next(action)
  if (action.type === UPDATE_BLOCK_SIZE || action.type === END_LINKING) {
    const state = getState()
    const blocks = cloneBlocks(state.blocks)
    const links = getLinks(state.links).map(({ fromId, toId }) => ({
      source: fromId,
      target: toId,
    }))
    simulation
      .nodes(blocks)
      .force('link', forceLink(links).id(({ id }: Block) => id))
      .alpha(1)
    computeStaticLayout()
    setTimeout(() => dispatch(updateForce(blocks)))
  }
  return result
}

export default force
