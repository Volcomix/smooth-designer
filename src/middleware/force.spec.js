//@flow
import force, { radius } from './force'
import { deleteBlock, updateBlockSize } from '../actions/blocks'
import { endLinking } from '../actions/links'

jest.useFakeTimers()
const mathMock: any = Math
mathMock.random = jest.fn(() => 0.3087575784346488)

const create = () => {
  const state = {
    blocks: {
      '0': { id: '0', name: 'Block 1', x: 0, y: 0, width: 10, height: 20 },
      '5': { id: '5', name: 'Block 2', x: 0, y: 0, width: 30, height: 40 },
    },
    links: { links: { '0': { id: '0', fromId: '0', toId: '5' } } },
  }
  const store = {
    getState: jest.fn(() => state),
    dispatch: jest.fn(),
  }
  const next = jest.fn()
  const invoke = action => force(store)(next)(action)
  return { store, next, invoke }
}

it('computes a block radius', () => {
  expect(
    radius({
      id: '0',
      name: 'Block',
      x: 0,
      y: 0,
      width: 10 * Math.cos(Math.PI / 3),
      height: 10 * Math.sin(Math.PI / 3),
    }),
  ).toEqual(5)
})

it('does not simulate force if not needed', () => {
  const { store, next, invoke } = create()
  const action = { type: 'DO_NOTHING' }
  invoke(action)
  expect(next).toHaveBeenCalledWith(action)
  jest.runAllTimers()
  expect(store.dispatch).not.toHaveBeenCalled()
})

it('updates positions when a block is deleted', () => {
  const { store, next, invoke } = create()
  const { dispatch, getState } = store
  const action = deleteBlock('5')
  invoke(action)
  expect(next).toHaveBeenCalledWith(action)
  expect(dispatch).not.toHaveBeenCalled()
  jest.runAllTimers()
  expect(dispatch.mock.calls).toMatchSnapshot()
  expect(getState()).toMatchObject({
    blocks: { '0': { x: 0, y: 0 }, '5': { x: 0, y: 0 } },
  })
})

it('updates positions when a block is sized', () => {
  const { store, next, invoke } = create()
  const { dispatch, getState } = store
  const action = updateBlockSize('5', 50, 60)
  invoke(action)
  expect(next).toHaveBeenCalledWith(action)
  expect(dispatch).not.toHaveBeenCalled()
  jest.runAllTimers()
  expect(dispatch.mock.calls).toMatchSnapshot()
  expect(getState()).toMatchObject({
    blocks: { '0': { x: 0, y: 0 }, '5': { x: 0, y: 0 } },
  })
})

it('updates positions when a link is added', () => {
  const { store, next, invoke } = create()
  const { dispatch, getState } = store
  const action = endLinking('5')
  invoke(action)
  expect(next).toHaveBeenCalledWith(action)
  expect(dispatch).not.toHaveBeenCalled()
  jest.runAllTimers()
  expect(dispatch.mock.calls).toMatchSnapshot()
  expect(getState()).toMatchObject({
    blocks: { '0': { x: 0, y: 0 }, '5': { x: 0, y: 0 } },
  })
})
