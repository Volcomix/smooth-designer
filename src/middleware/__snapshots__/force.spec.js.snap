// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with blocks updates positions when a block is added 1`] = `
Array [
  Array [
    Object {
      "blocks": Array [
        Object {
          "height": 0,
          "id": "0",
          "index": 0,
          "name": "Block",
          "vx": -0.000320572368372346,
          "vy": 6.015067608080946e-12,
          "width": 0,
          "x": -150.55847784868254,
          "y": -5.636344839385168e-8,
        },
        Object {
          "height": 0,
          "id": "999",
          "index": 1,
          "name": "New block",
          "vx": 0.000320572368372346,
          "vy": -6.015067608080946e-12,
          "width": 0,
          "x": 150.55847784868254,
          "y": 5.636344839385168e-8,
        },
      ],
      "type": "UPDATE_FORCE",
    },
  ],
]
`;

exports[`without blocks updates position when a block is added 1`] = `
Array [
  Array [
    Object {
      "blocks": Array [
        Object {
          "height": 0,
          "id": "999",
          "index": 0,
          "name": "New block",
          "vx": -1.27321588949404e-7,
          "vy": 0,
          "width": 0,
          "x": 0.0009098280429729471,
          "y": 0,
        },
      ],
      "type": "UPDATE_FORCE",
    },
  ],
]
`;
